Specify the HELP command. (#132)
ircdocs/modern-irc: _includes/modern-appendix.md

Indicates that a {% message MODE %} command affecting a user failed because they were trying to set or view modes for other users. The text used in the last param of this message varies, for instance when trying to view modes for another user, a server may send: `"Can't view modes for other users"`.

{% numericheader ERR_HELPNOTFOUND %}

      "<client> <subject> :No help available on this topic"

Indicates that a {% message HELP %} command requested help on a subject the server does not know about.

> The `<subject>` MUST be the one requested by the client, up to casing; unless it would be an invalid parameter, in which case it MUST be `*`.

Reviewer: The `up to casing` should be changed to match the new text we decide on below.

> The `<subject>` MUST be the one requested by the client, up to casing; unless it would be an invalid parameter, in which case it MUST be `*`.

Reviewer: ```suggestion
The `<subject>` MUST be the one requested by the client, but may be casefolded; unless it would be an invalid parameter, in which case it MUST be `*`.
```

Indicates that there was a problem with a mode parameter. Replaces various implementation-specific mode-specific numerics.

{% numericheader RPL_HELPSTART %}

    "<client> <subject> :<first line of help section>"

Indicates the start of a reply to a {% command HELP %} command.

> The text used in the last param of this message may vary, and SHOULD be displayed as-is by IRC clients to their users; possibly emphasized as a the title of the help section.

Reviewer 2: ```suggestion
The text used in the last parameter of this message may vary, and SHOULD be displayed as-is by IRC clients to their users; possibly emphasized as the title of the help section.
```

The `<subject>` MUST be the one requested by the client, but may be casefolded; unless it would be an invalid parameter, in which case it MUST be `*`.

{% numericheader RPL_HELPTXT %}

    "<client> <subject> :<line of help text>"

Returns a line of {% command HELP %} text to the client. Lines MAY be wrapped to a certain line length by the server. Note that the final line MUST be a {% numeric RPL_ENDOFHELP %} numeric.

> The `<subject>` MUST be the one requested by the client, up to casing; unless it would be an invalid parameter, in which case it MUST be `*`.
> 
> {% numericheader RPL_ENDOFHELP %}
> 
>     "<client> <subject> :<last line of help text>"
> 
> Returns the final {% command HELP %} line to the client.
> 
> The `<subject>` MUST be the one requested by the client, up to casing; unless it would be an invalid parameter, in which case it MUST be `*`.

Reviewer: ```suggestion
The `<subject>` MUST be the one requested by the client, but may be casefolded; unless it would be an invalid parameter, in which case it MUST be `*`.

{% numericheader RPL_ENDOFHELP %}

    "<client> <subject> :<last line of help text>"

Returns the final {% command HELP %} line to the client.

The `<subject>` MUST be the one requested by the client, but may be casefolded; unless it would be an invalid parameter, in which case it MUST be `*`.
```

Indicates that there was a problem with a mode parameter. Replaces various implementation-specific mode-specific numerics.

{% numericheader RPL_HELPSTART %}

>     "<client> <subject> :<firt line of help section>"
> 
> Indicates the start of a reply to a {% command HELP %} command to the client.
> The text used in the last param of this message may vary, and SHOULD be displayed as-is by IRC clients to their users; possibly emphasized as a the title of the help section.
> 
> The `<subject>` MUST be the one requested by the client, up to casing; unless it would be an invalid parameter, in which case it MUST be `*`.

Reviewer: Slight spelling fix and clarification of the `up to casing` part. If this new text is wrong, replace `but may be casefolded` with `case sensitive` instead.

The two other `up to casing` parts below should also be replaced with the new text.

```suggestion
    "<client> <subject> :<first line of help section>"

Indicates the start of a reply to a {% command HELP %} command.
The text used in the last param of this message may vary, and SHOULD be displayed as-is by IRC clients to their users; possibly emphasized as a the title of the help section.

The `<subject>` MUST be the one requested by the client, but may be casefolded; unless it would be an invalid parameter, in which case it MUST be `*`.
```

{% numericheader RPL_HELPTXT %}

    "<client> <subject> :<line of help text>"

> When sending a reply to a {% command HELP %} command to the client, servers reply with each line of the help section as this numeric. Help lines MAY be wrapped to 80 characters by the server.

Reviewer: Simplified text and changed the wrapping slightly. I think ours wraps at like 90 characters or something, makes sense to let servers choose how wide to wrap there.

```suggestion
Returns a line of {% command HELP %} text to the client. The text MUST be shown to the user as-is, and lines MAY be wrapped to a certain line length by the server.
```

Author: That was copy-pasted from RPL_MOTD's description

Author: Why did you make this a MUST? I don't think it's appropriate here, it doesn't break the protocol to transform it.

Reviewer: Ah, hm. I'll toss in a PR updating `RPL_MOTD`'s description to match this, how's that sound?

Author: ok

Reviewer: I think clients just show the help text as is, and especially since the help text can describe server-specific differences or notes it doesn't make sense to really let clients transform it. I'd be fine having it as a SHOULD or a MUST though, which would you prefer?

Reviewer: (oh wait, do you mean 'transform' as in change whitespace or like replace messages with translated versions or something?)

Author: or interpret color codes, convert it to HTML or whatever. It doesn't matter to the protocol. See https://datatracker.ietf.org/doc/html/rfc2119#section-6

Reviewer: ah, alright that makes sense. I was expecting like cases where clients would be substituting the actual text for their own (not applicable) text instead. give me a sec

Author: Hmm then it would be an issue. For example in weechat, `/help` shows the internal help instead of sending `HELP` to the server

Reviewer: updated!

The `<subject>` MUST be the one requested by the client, up to casing; unless it would be an invalid parameter, in which case it MUST be `*`.

{% numericheader RPL_ENDOFHELP %}

    "<client> <subject> :<last line of help text>"

> Indicates the end of the reply to the {% command HELP %} command to the client.
> The text used in the last param of this message may vary, and SHOULD be shown to the user as any other line.

Reviewer: I don't think we need to explain this as much, that's covered by the examples in the linked `HELP` command. I think this matches the new text for `RPL_HELPTXT` pretty well and explains enough to get the point across.

```suggestion
Returns the final {% command HELP %} line to the client.
The text MUST be shown to the user as-is.
```

Author: same comment on the MUST

Reviewer: removed the MUST!
